hcloud/resource_hcloud_reversedns.go

```
@@ -0,0 +1,208 @@
+	}
+	switch {
+	case floatingIP != nil:
+		dnsPtr := floatingIP.DNSPtrForIP(ip)
+		if dnsPtr != "" {
+			d.Set("dnsPtr", dnsPtr)
+			d.Set("floating_ip_id", floatingIP.ID)
+			d.Set("ip_address", ipAddress)
+			d.SetId(generateRDNSID(nil, floatingIP, ipAddress))
+		} else {
+			d.SetId("")
+		}
+	case server != nil:
+		if ip.To4() != nil {
+			d.SetId(generateRDNSID(server, nil, ipAddress))
+			d.Set("dnsPtr", server.PublicNet.IPv4.DNSPtr)
+			d.Set("server_id", server.ID)
+			d.Set("ip_address", server.PublicNet.IPv4.IP)
+		} else if ip.To16() != nil {
+			for rdns := range server.PublicNet.IPv6.DNSPtr {
+				if rdns == ipAddress {
+					d.SetId(generateRDNSID(server, nil, ipAddress))
+					d.Set("dnsPtr", server.PublicNet.IPv6.DNSPtrForIP(ip))
+					d.Set("ip_address", ipAddress)
+					d.Set("server_id", server.ID)
+				}
+			}
+		}
+
+	}
+	return nil
+}
+
+func resourceReverseDNSCreate(d *schema.ResourceData, m interface{}) error {
+	client := m.(*hcloud.Client)
+	ctx := context.Background()
+	ip := d.Get("ip_address").(string)
+	ptr := d.Get("dns_ptr").(string)
+
+	rdnsType := ""
+
+	id, ok := d.GetOk("server_id")
+	if ok == false {
+		id, ok = d.GetOk("floating_ip_id")
+		switch {
+		case !ok:
+			log.Printf("[WARN] Invalid id (%s), removing from state: %v", d.Id(), ok)
+			d.SetId("")
+			return nil
+		}
+		floatingIP, _, _ := client.FloatingIP.GetByID(ctx, id.(int))
+		d.SetId(generateRDNSID(nil, floatingIP, ip))
+		rdnsType = "floatingIP"
+	} else {
+		server, _, _ := client.Server.GetByID(ctx, id.(int))
+		d.SetId(generateRDNSID(server, nil, ip))
+		rdnsType = "server"
+	}
+
+	switch {
+	case rdnsType == "floating_ip":
+		floatingIP, _, err := client.FloatingIP.GetByID(ctx, id.(int))
+		if err != nil {
+			return err
+		}
+		if floatingIP == nil {
+			log.Printf("[WARN] Floating IP (%s) not found, removing from state", d.Id())
+			d.SetId("")
+			return nil
+		}
+		action, _, err := client.FloatingIP.ChangeDNSPtr(ctx, floatingIP, ip, &ptr)
```

err is never checked